core/bridges/orm.go

```diff
@@ -179,26 +185,43 @@ func (o *orm) UpdateBridgeType(ctx context.Context, bt *BridgeType, btr *BridgeT
 	return err
 }
 
-func (o *orm) GetCachedResponse(ctx context.Context, dotId string, specId int32, maxElapsed time.Duration) (response []byte, err error) {
+func (o *orm) GetCachedResponse(ctx context.Context, dotId string, specId int32, maxElapsed time.Duration) ([]byte, error) {
+	// prefer to get latest value from cache
+	cachedResponse, inCache := o.bridgeLastValueCache.Load(responseKey(dotId, specId))
+	if inCache {
+		return cachedResponse.([]byte), nil
+	}
+
+	// if not in cache, fallback to the data store
 	stalenessThreshold := time.Now().Add(-maxElapsed)
 	sql := `SELECT value FROM bridge_last_value WHERE
 				dot_id = $1 AND 
 				spec_id = $2 AND 
 				finished_at > ($3)	
 				ORDER BY finished_at 
 				DESC LIMIT 1;`
-	err = pkgerrors.Wrap(o.ds.GetContext(ctx, &response, sql, dotId, specId, stalenessThreshold), fmt.Sprintf("failed to fetch last good value for task %s spec %d", dotId, specId))
-	return
+
+	var response []byte
+
+	if err := pkgerrors.Wrap(
+		o.ds.GetContext(ctx, &response, sql, dotId, specId, stalenessThreshold),
+		fmt.Sprintf("failed to fetch last good value for task %s spec %d", dotId, specId),
+	); err != nil {
+		return nil, err
+	}
+
+	o.bridgeLastValueCache.Store(responseKey(dotId, specId), response)
+
+	return response, nil
 }
 
 func (o *orm) UpsertBridgeResponse(ctx context.Context, dotId string, specId int32, response []byte) error {
-	sql := `INSERT INTO bridge_last_value(dot_id, spec_id, value, finished_at) 
-				VALUES($1, $2, $3, $4)
-			ON CONFLICT ON CONSTRAINT bridge_last_value_pkey
-				DO UPDATE SET value = $3, finished_at = $4;`
+	o.bridgeLastValueCache.Store(responseKey(dotId, specId), response)
 
-	_, err := o.ds.ExecContext(ctx, sql, dotId, specId, response, time.Now())
-	return pkgerrors.Wrap(err, "failed to upsert bridge response")
+	// async call to the data store with retry
+	go o.upsertResponse(ctx, dotId, specId, response, UpsertRetryDelay)
```

Comment:
Problem here is that we will try to open an additional DB connection for every stored result, and the number of connections is starting to become an issue. Could we consider instead feeding responses into a queue which is pulled from on a loop that does multiple inserts at a time, so it only consumes one connection?

You could also optimize so that only the latest for every key is persisted.